File: java-automation/src/main/java/com/automation/factory/PlaywrightFactory.java
package com.automation.factory;

import com.microsoft.playwright.*;

/**
 * Thread-safe Playwright lifecycle manager.
 *
 * <p>State is held in {@link ThreadLocal}s, so each test thread owns an isolated
 * Playwright / Browser / Context / Page. This lets the suite run in parallel
 * (JUnit Platform parallel execution) without threads clobbering one another.
 *
 * <p>Headless is the default (CI-friendly); set {@code HEADLESS=false} to watch
 * the browser locally.
 */
public class PlaywrightFactory {
    private static final ThreadLocal<Playwright> PLAYWRIGHT = new ThreadLocal<>();
    private static final ThreadLocal<Browser> BROWSER = new ThreadLocal<>();
    private static final ThreadLocal<BrowserContext> CONTEXT = new ThreadLocal<>();
    private static final ThreadLocal<Page> PAGE = new ThreadLocal<>();

    public static Page createPage() {
        boolean headless = !"false".equalsIgnoreCase(System.getenv("HEADLESS"));

        PLAYWRIGHT.set(Playwright.create());
        BROWSER.set(PLAYWRIGHT.get().chromium()
            .launch(new BrowserType.LaunchOptions().setHeadless(headless)));
        CONTEXT.set(BROWSER.get().newContext());

        Page page = CONTEXT.get().newPage();
        page.setDefaultTimeout(30000);
        PAGE.set(page);
        return page;
    }

    /** Current thread's page (used by the screenshot-on-failure extension). */
    public static Page getPage() {
        return PAGE.get();
    }

    public static void closeAll() {
        if (CONTEXT.get() != null) {
            CONTEXT.get().close();
            CONTEXT.remove();
        }
        if (BROWSER.get() != null) {
            BROWSER.get().close();
            BROWSER.remove();
        }
        if (PLAYWRIGHT.get() != null) {
            PLAYWRIGHT.get().close();
            PLAYWRIGHT.remove();
        }
        PAGE.remove();
    }
}

File: python-automation/src/automation/factory/playwright_factory.py
from typing import Optional

from playwright.sync_api import (
    sync_playwright,
    Playwright,
    Browser,
    BrowserContext,
    Page,
)


class PlaywrightFactory:
    """
    Instance-based Playwright lifecycle manager.

    Each instance owns its own Playwright / Browser / Context / Page, so every
    test gets a fully isolated browser session. This makes the suite safe to run
    in parallel (e.g. ``pytest -n auto`` with pytest-xdist) — there is no shared
    static state to clobber between workers.
    """

    def __init__(self) -> None:
        self._playwright: Optional[Playwright] = None
        self._browser: Optional[Browser] = None
        self._context: Optional[BrowserContext] = None
        self._page: Optional[Page] = None

    def create_page(self, headless: bool = True, slow_mo: int = 0) -> Page:
        """Start Playwright and return a fresh, isolated page."""
        self._playwright = sync_playwright().start()
        self._browser = self._playwright.chromium.launch(
            headless=headless,
            slow_mo=slow_mo,
            args=["--disable-blink-features=AutomationControlled"],
        )
        self._context = self._browser.new_context()
        self._page = self._context.new_page()
        self._page.set_default_timeout(30000)
        return self._page

    @property
    def page(self) -> Optional[Page]:
        return self._page

    def close_all(self) -> None:
        """Tear down all resources owned by this instance."""
        if self._context is not None:
            self._context.close()
            self._context = None
        if self._browser is not None:
            self._browser.close()
            self._browser = None
        if self._playwright is not None:
            self._playwright.stop()
            self._playwright = None
        self._page = None
